drivers/periph_common/timer.c

```diff
@@ -38,4 +38,53 @@ uword_t timer_query_channel_numof(tim_t dev)
     (void)dev;
     return TIMER_CHANNEL_NUMOF;
 }
+
+uint32_t timer_get_closest_freq(tim_t dev, uint32_t target)
+{
+    uint32_t freq;
+    uword_t idx_max = timer_query_freqs_numof(dev) - 1;
+
+    /* use binary search to find one of the three possibilities:
+     * 1. If an exact match is possible: The exact match is found
+     * 2. Otherwise:
+     *      a) the highest frequency below the target, or
+     *      b) the lowest frequency above the target
+     */
+    uword_t idx_lower = 0;
+    uword_t idx_upper = idx_max;
+    while (idx_lower != idx_upper) {
+        uword_t idx_mid = (idx_lower + idx_upper) >> 1;
+        freq = timer_query_freqs(dev, idx_mid);
+        if (freq > target) {
+            idx_lower = idx_mid + 1;
+        }
+        else {
+            idx_upper = idx_mid;
+        }
+    }
+
+    freq = timer_query_freqs(dev, idx_lower);
+    if ((freq < target) && (idx_lower > 0)) {
+        /* binary search yielded the closest frequency below the target. But
+         * maybe the closest frequency above the target is actually better. */
+        uint32_t diff = target - freq;
+        uint32_t alternative = timer_query_freqs(dev, idx_lower - 1);
+        if (target + diff > alternative) {
+            /* the alternative is better */
+            return alternative;
+        }
+    }
+    else if ((freq > target) && (idx_lower < idx_max)) {
+        /* Got a frequency above the target, maybe the one below would be
+         * a closer match */
+        uint32_t diff = target - freq;
```

Comment:
```suggestion
        uint32_t diff = freq - target;
```

I'd say? :thinking: Might be worth somehow catching such wrong behavior in a test case.

Reply:
This would probably still work because the uint32_t wraps around in line 80 and then again in 82. But certainly, this is not good behavior 😅